### PR TITLE
Option to change Context options

### PR DESCRIPTION
**Pull request by btepu002xamk**
User can change context options

### CODE REVIEW

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 1]
```diff
@@ -27,6 +27,9 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
 }) => {
   const [ metaformSectionTitle, setMetaformSectionTitle ] = useState<string | undefined>("");
   const [ columnType, setColumnType ] = useState<string>("");
+  const [ contextFormOption, setContextFormOption ] = React.useState<boolean>(false);
+  const [ contextManagementOption, setContextManagementOption ] = React.useState<boolean>(false);
+  const [ contextManagementListOption, setContextManagementListOption ] = React.useState<boolean>(false);
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:32]:
```suggestion
  const [ contextOption, setContextOption ] = useState<FormContext[]>( [FormContext.FORM, FormContext.MANAGEMENT ]);
```

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 2]
```diff
@@ -212,6 +250,73 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
...
+                checked={ contextManagementListOption }
+                onChange={ event => updateContexts("MANAGEMENT_LIST", event.target.checked) }
+              />
+            }
+          />
+        </FormControl>
+      </>
+    );
+  };
   /**
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:320]:
```suggestion
  const renderContextOptions = () => (
      <>
        <Divider/>
        <Typography variant="subtitle1">{ strings.draftEditorScreen.editor.features.ContextVisibilityInfo }</Typography>
        <FormControl>
          <FormControlLabel
            label={ strings.draftEditorScreen.editor.features.contextFormVisibility }
            control={
              <Checkbox
                checked={ contextFormOption }
                onChange={ event => updateContexts("FORM", event.target.checked) }
              />
            }
          />
        </FormControl>
        <FormControl>
          <FormControlLabel
            label={ strings.draftEditorScreen.editor.features.contextManagementVisibility }
            control={
              <Checkbox
                checked={ contextManagementOption }
                onChange={ event => updateContexts("MANAGEMENT", event.target.checked) }
              />
            }
          />
        </FormControl>
        <FormControl>
          <FormControlLabel
            label={ strings.draftEditorScreen.editor.features.contextManagementListVisibility }
            control={
              <Checkbox
                checked={ contextManagementListOption }
                onChange={ event => updateContexts("MANAGEMENT_LIST", event.target.checked) }
              />
            }
          />
        </FormControl>
      </>
    );
  );
  
  /**
```

Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:320]:
```suggestion
  const renderContextOptions = () => (
      <>
        <Divider/>
        <Typography variant="subtitle1">{ strings.draftEditorScreen.editor.features.ContextVisibilityInfo }</Typography>
        <FormControl>
          <FormControlLabel
            label={ strings.draftEditorScreen.editor.features.contextFormVisibility }
            control={
              <Checkbox
                checked={ contextFormOption }
                onChange={ event => updateContexts("FORM", event.target.checked) }
              />
            }
          />
        </FormControl>
        <FormControl>
          <FormControlLabel
            label={ strings.draftEditorScreen.editor.features.contextManagementVisibility }
            control={
              <Checkbox
                checked={ contextManagementOption }
                onChange={ event => updateContexts("MANAGEMENT", event.target.checked) }
              />
            }
          />
        </FormControl>
        <FormControl>
          <FormControlLabel
            label={ strings.draftEditorScreen.editor.features.contextManagementListVisibility }
            control={
              <Checkbox
                checked={ contextManagementListOption }
                onChange={ event => updateContexts("MANAGEMENT_LIST", event.target.checked) }
              />
            }
          />
        </FormControl>
      </>
    );
  );
  
  /**
```

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 3]
```diff
@@ -212,6 +250,73 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
...
+        return null;
+      });
+    }
+  };
+
+  /**
+   * Render contexts options
+   */
+  const renderContextOptions = () => {
+    return (
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:280]:
```suggestion
  const renderContextOptions = () => (
```

File: src/components/editor/metaform-editor-right-drawer-feature.tsx (hunk 2 above)
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:320]:
```suggestion
  );
  
  /**
```

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 4]
```diff
@@ -212,6 +250,73 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
...
+      <>
+        <Divider/>
+        <Typography variant="subtitle1">{ strings.draftEditorScreen.editor.features.ContextVisibilityInfo }</Typography>
+        <FormControl>
+          <FormControlLabel
+            label={ strings.draftEditorScreen.editor.features.contextFormVisibility }
+            control={
+              <Checkbox
+                checked={ contextFormOption }
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:289]:
```suggestion
                checked={ contextOption.includes(FormContext.FORM) }
```

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 5]
```diff
@@ -212,6 +250,73 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
...
+              />
+            }
+          />
+        </FormControl>
+        <FormControl>
+          <FormControlLabel
+            label={ strings.draftEditorScreen.editor.features.contextManagementVisibility }
+            control={
+              <Checkbox
+                checked={ contextManagementOption }
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:300]:
```suggestion
                checked={ contextOption.includes(FormContext.MANAGEMENT) }
```

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 6]
```diff
@@ -212,6 +250,73 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
...
+              />
+            }
+          />
+        </FormControl>
+        <FormControl>
+          <FormControlLabel
+            label={ strings.draftEditorScreen.editor.features.contextManagementListVisibility }
+            control={
+              <Checkbox
+                checked={ contextManagementListOption }
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:311]:
```suggestion
                checked={ contextOption.includes(FormContext.MANAGEMENT_LIST) }
```

**Review by villejuutila — COMMENTED** [review]
Added comments

File: src/types/index.ts [hunk 7]
```diff
@@ -122,6 +122,15 @@ export enum SosmetaType {
   STRING = "string"
 }
 
+/**
+ * Enum for different types of Contexts
```
Comment (villejuutila) [inline src/types/index.ts:126]:
```suggestion
 * Enum for different types of Contexts for form fields
```

File: src/localization/fi.json [hunk 8]
```diff
@@ -240,7 +240,11 @@
         "columnNumberType": "Numerokenttä",
         "textOfSubmitButton": "Painikkeen teksti",
         "newFieldOption":"Valintakenttä",
-        "addCustomHtml":"Lisää tähän HTML-sisältöä"
+        "addCustomHtml":"Lisää tähän HTML-sisältöä",
+        "contextFormVisibility":"Näkyy lomakkeella täyttäessä",
+        "contextManagementVisibility":"Näkyy hallinnointi näkymässä",
+        "contextManagementListVisibility":"Näkyy hallinnointilista näkymässä",
```
Comment (villejuutila) [inline src/localization/fi.json:246]:
```suggestion
        "contextManagementListVisibility":"Näkyy hallinnointi listaus näkymässä",
```

File: src/localization/fi.json [hunk 9]
```diff
@@ -240,7 +240,11 @@
         "columnNumberType": "Numerokenttä",
         "textOfSubmitButton": "Painikkeen teksti",
         "newFieldOption":"Valintakenttä",
-        "addCustomHtml":"Lisää tähän HTML-sisältöä"
+        "addCustomHtml":"Lisää tähän HTML-sisältöä",
+        "contextFormVisibility":"Näkyy lomakkeella täyttäessä",
```
Comment (villejuutila) [inline src/localization/fi.json:244]:
```suggestion
        "contextFormVisibility":"Näkyy lomaketta täyttäessä",
```

File: src/localization/en.json [hunk 10]
```diff
@@ -242,7 +242,11 @@
         "columnNumberType": "Number field",
         "textOfSubmitButton": "Button text",
         "newFieldOption":"Option field",
-        "addCustomHtml":"Add html code in field"
+        "addCustomHtml":"Add HTML code in field",
+        "contextFormVisibility":"Shown on the form for customer",
+        "contextManagementVisibility":"Shown on management screen",
+        "contextManagementListVisibility":"Shown on managementlist screen",
```
Comment (villejuutila) [inline src/localization/en.json:248]:
```suggestion
        "contextManagementListVisibility":"Shown on management list screen",
```

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 11]
```diff
@@ -554,6 +647,12 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
     );
   };
 
+  useEffect(() => {
+    if (fieldIndex !== undefined) {
+      checkContextSettings();
+    }
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:653]: Move this conditional check into _checkContextSettings()_ method and only call the method itself from useEffect.

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 12]
```diff
@@ -27,7 +27,7 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
 }) => {
   const [ metaformSectionTitle, setMetaformSectionTitle ] = useState<string | undefined>("");
   const [ columnType, setColumnType ] = useState<string>("");
-
+  const [ contextOption, setContextOption ] = useState<FormContext[]>([FormContext.FORM, FormContext.MANAGEMENT]);
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:30]:
```suggestion
  const [ contextOption, setContextOption ] = useState<FormContext[]>([]);
```

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 13]
```diff
@@ -197,6 +197,34 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
...
+          return null;
+        });
+      }
+      if (checked) {
+        const updatedForm = produce(pendingForm, draftForm => {
+          draftForm.sections?.[sectionIndex]?.fields?.[fieldIndex]?.contexts?.push(selectedContext);
+        });
+        setPendingForm(updatedForm);
+      }
+    }
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:225]: Move `if (checked)` into else branch of `if (!checked)`

File: src/utils/metaform-utils.ts [hunk 14]
```diff
@@ -73,94 +73,95 @@ namespace MetaformUtils {
...
+    switch (fieldType) {
+      case MetaformFieldType.Select:
+      case MetaformFieldType.Radio:
+      case MetaformFieldType.Checklist:
+        return {
+          name: name ?? fieldType,
+          title: title ?? fieldType,
+          type: fieldType,
+          required: required ?? false,
+          contexts: ["FORM", "MANAGEMENT"],
```
Comment (villejuutila) [inline src/utils/metaform-utils.ts:85]:
```suggestion
          contexts: [ FormContext.FORM, FormContext.MANAGEMENT ]
```
Use FormContext enum values instead.
Do the same thing for all cases in this switch.

**Review by villejuutila — CHANGES_REQUESTED** [review]
Added comments

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 15]
```diff
@@ -197,6 +197,33 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
     setPendingForm(updatedForm);
   };
 
+  /**
+   * Update contexts of field
+   * 
+   * @param selectedContext Selected context Option
+   * @param checked Is context option checked or not
+   */
+  const updateContexts = (selectedContext: string, checked: boolean) => {
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:206]:
```suggestion
  const updateContexts = (selectedContext: FormContext, checked: boolean) => {
```

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 16]
```diff
@@ -197,6 +197,33 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
...
+    if (sectionIndex !== undefined && fieldIndex !== undefined) {
+      if (!checked) {
+        pendingForm.sections![sectionIndex].fields![fieldIndex].contexts!.map((context, index) => {
+          if (selectedContext === context) {
+            const updatedForm = produce(pendingForm, draftForm => {
+              draftForm.sections?.[sectionIndex]?.fields?.[fieldIndex]?.contexts?.splice(index, 1);
+            });
+            setPendingForm(updatedForm);
+          }
+          return null;
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:216]: Isn't this return redundant? Change .map to .forEach

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 17]
```diff
@@ -212,6 +239,71 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
...
+          case FormContext.FORM:
+            return setContextOption(contextOptionList => [...contextOptionList, context]);
+          case FormContext.MANAGEMENT:
+            return setContextOption(contextOptionList => [...contextOptionList, context]);
+          case FormContext.MANAGEMENT_LIST:
+            return setContextOption(contextOptionList => [...contextOptionList, context]);
+          default:
+            break;
+        }
+        return null;
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:259]: Isn't this return redundant? Change .map to .forEach

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 18]
```diff
@@ -212,6 +239,71 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
     setPendingForm(updatedForm);
   };
 
+  /**
+   * When changing field check what context settings field have
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:243]:
```suggestion
   * Checks fields contexts
```

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 19]
```diff
@@ -212,6 +239,71 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
...
+        return null;
+      });
+    }
+  };
+
+  /**
+   * Render contexts options
+   */
+  const renderContextOptions = () => (
+    <>
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:268]:
```suggestion
    <Stack spacing={ 2 }>
```

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 20]
```diff
@@ -212,6 +239,71 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
...
+          label={ strings.draftEditorScreen.editor.features.contextManagementListVisibility }
+          control={
+            <Checkbox
+              checked={ contextOption.includes(FormContext.MANAGEMENT_LIST) }
+              onChange={ event => updateContexts(FormContext.MANAGEMENT_LIST, event.target.checked) }
+            />
+          }
+        />
+      </FormControl>
+    </>
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:304]:
```suggestion
    </Stack>
```

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 21]
```diff
@@ -405,25 +497,19 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
...
+        case MetaformFieldType.Checklist:
+        case MetaformFieldType.Radio:
+        case MetaformFieldType.Select:
+          return renderFieldOptions();
+        case MetaformFieldType.Html:
+          return renderHtmlEditor(selectedField);
+        case MetaformFieldType.Table:
+          return renderTableColumnFeatures();
+        default:
+          return null;
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:512]:
```suggestion
          break;
```

File: src/localization/fi.json [hunk 22]
```diff
@@ -240,7 +240,11 @@
         "columnNumberType": "Numerokenttä",
         "textOfSubmitButton": "Painikkeen teksti",
         "newFieldOption":"Valintakenttä",
-        "addCustomHtml":"Lisää tähän HTML-sisältöä"
+        "addCustomHtml":"Lisää tähän HTML-sisältöä",
+        "contextFormVisibility":"Näkyy lomaketta täyttäessä",
+        "contextManagementVisibility":"Näkyy hallinnointi näkymässä",
+        "contextManagementListVisibility":"Näkyy hallinnointilista näkymässä",
```
Comment (villejuutila) [inline src/localization/fi.json:246]:
```suggestion
        "contextManagementListVisibility":"Näkyy hallinnoin listaus näkymässä",
```

File: src/localization/fi.json [hunk 23]
```diff
@@ -240,7 +240,11 @@
...
+        "ContextVisibilityInfo":"Kentän hallinnointi asetukset"
```
Comment (villejuutila) [inline src/localization/fi.json:247]:
```suggestion
        "contextVisibilityInfo": "Kentän hallinnointi asetukset"
```

File: src/localization/fi.json [hunk 24]
```diff
@@ -240,7 +240,11 @@
         "columnNumberType": "Numerokenttä",
         "textOfSubmitButton": "Painikkeen teksti",
         "newFieldOption":"Valintakenttä",
-        "addCustomHtml":"Lisää tähän HTML-sisältöä"
+        "addCustomHtml":"Lisää tähän HTML-sisältöä",
```
Comment (villejuutila) [inline src/localization/fi.json:243]:
```suggestion
        "addCustomHtml": "Lisää tähän HTML-sisältöä",
```

File: src/localization/fi.json [hunk 25]
```diff
@@ -240,7 +240,11 @@
...
+        "contextFormVisibility":"Näkyy lomaketta täyttäessä",
```
Comment (villejuutila) [inline src/localization/fi.json:244]:
```suggestion
        "contextFormVisibility": "Näkyy lomaketta täyttäessä",
```

File: src/localization/fi.json [hunk 26]
```diff
@@ -240,7 +240,11 @@
...
+        "contextManagementVisibility":"Näkyy hallinnointi näkymässä",
```
Comment (villejuutila) [inline src/localization/fi.json:245]:
```suggestion
        "contextManagementVisibility": "Näkyy hallinnointi näkymässä",
```

File: src/localization/fi.json (hunk 22 above)
Comment (villejuutila) [inline src/localization/fi.json:246]:
```suggestion
        "contextManagementListVisibility": "Näkyy hallinnointilista näkymässä",
```

File: src/types/index.ts [hunk 27]
```diff
@@ -122,6 +122,15 @@ export enum SosmetaType {
   STRING = "string"
 }
 
+/**
+ * Enum for different types of Contexts for form fields
```
Comment (villejuutila) [inline src/types/index.ts:126]:
```suggestion
 * Enum for MetaformField context values
```

**Review by villejuutila — CHANGES_REQUESTED** [review]
Added comments

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 28]
```diff
@@ -212,6 +238,70 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
...
+        switch (context) {
+          case FormContext.FORM:
+            return setContextOption(contextOptionList => [...contextOptionList, context]);
+          case FormContext.MANAGEMENT:
+            return setContextOption(contextOptionList => [...contextOptionList, context]);
+          case FormContext.MANAGEMENT_LIST:
+            return setContextOption(contextOptionList => [...contextOptionList, context]);
+          default:
+            break;
+        }
```
Comment (Chi-Chu319) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:257]: Whats the point since they are all the same operation

**Review by Chi-Chu319 — APPROVED** [review]

**Review by villejuutila — APPROVED** [review]
Approved

**Pull request merged**